### PR TITLE
overlord/ifacestate,daemon: setup security on conect and disconnect

### PR DESCRIPTION
**Pull request by zyga**
This branch makes connect/disconnect setup security of both affected snaps.

### CODE REVIEW

File: daemon/api_test.go [hunk 1]
```diff
@@ -108,10 +110,15 @@ func (s *apiSuite) SetUpTest(c *check.C) {
 		configs: map[string]string{},
 	}
 	s.d = nil
+	// Disable real security backends for all API tests
+	s.restoreBackends = ifacestate.MockSecurityBackendsForSnap(
```
Comment (niemeyer) [inline daemon/api_test.go]: How about making that simply `MockSecurityBackends(nil)`?  I don't think any call site is using per-snap backends, which means all of that boilerplate is just unnecessary trouble.

Comment (zyga) [inline daemon/api_test.go]: Good idea, done.

File: overlord/ifacestate/ifacemgr.go [hunk 2]
```diff
@@ -423,3 +457,15 @@ func (m *InterfaceManager) Stop() {
 func (m *InterfaceManager) Repository() *interfaces.Repository {
 	return m.repo
 }
+
+// MockSecurityBackendsForSnap mocks the list of security backends that are used for setting up security.
+//
+// This function is public because it is referenced in the daemon
+func MockSecurityBackendsForSnap(fn func(snapInfo *snap.Info) []interfaces.SecurityBackend) func() {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: I actually meant this:
```
func MockSecurityBackends(backends []interfaces.SecurityBackend)
```

File: overlord/ifacestate/ifacemgr.go [hunk 3]
```diff
@@ -365,8 +365,25 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
...
 	plug := m.repo.Plug(plugRef.Snap, plugRef.Name)
+	slot := m.repo.Slot(slotRef.Snap, slotRef.Name)
+
+	for _, snapInfo := range []*snap.Info{plug.Snap, slot.Snap} {
+		var snapState snapstate.SnapState
+		if err := snapstate.Get(st, snapInfo.Name(), &snapState); err != nil {
+			task.Errorf("cannot get state of snap %q: %s", snapInfo.Name(), err)
+			return state.Retry
+		}
+		for _, backend := range securityBackendsForSnap(snapInfo) {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Similarly,
```
for _, backend := range securityBackends { ... }
```

File: overlord/ifacestate/ifacemgr.go (hunk 2 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Ahh, sorry. Will do in a sec.

Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 3 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go [hunk 4]
```diff
@@ -365,8 +365,25 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 	}
 
 	plug := m.repo.Plug(plugRef.Snap, plugRef.Name)
+	slot := m.repo.Slot(slotRef.Snap, slotRef.Name)
+
+	for _, snapInfo := range []*snap.Info{plug.Snap, slot.Snap} {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Can we please do this here instead:
```
err := setupSnapSecurity(task, plug.Snap, m.repo)
if err != nil {
        return err
}
err = setupSnapSecurity(task, slot.Snap, m.repo)
if err != nil {
        return err
}

File: overlord/ifacestate/ifacemgr.go [hunk 5]
```diff
@@ -390,6 +407,23 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	plug := m.repo.Plug(plugRef.Snap, plugRef.Name)
+	slot := m.repo.Slot(slotRef.Snap, slotRef.Name)
+
+	for _, snapInfo := range []*snap.Info{plug.Snap, slot.Snap} {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Same as above.

File: overlord/ifacestate/ifacemgr.go [hunk 6]
```diff
@@ -423,3 +457,13 @@ func (m *InterfaceManager) Stop() {
 func (m *InterfaceManager) Repository() *interfaces.Repository {
 	return m.repo
 }
+
+// MockSecurityBackends mocks the list of security backends that are used for setting up security.
+//
+// This function is public because it is referenced in the daemon
+func MockSecurityBackends(backends []interfaces.SecurityBackend) func() {
+	securityBackends = func(snapInfo *snap.Info) []interfaces.SecurityBackend {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: This should be
```
oldBackends = securityBackends
securityBackends = backends
return func() { securityBackends = oldBackends }
```

Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done now

File: overlord/ifacestate/ifacemgr.go (hunk 4 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 5 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go [hunk 7]
```diff
@@ -130,6 +130,22 @@ func (m *InterfaceManager) reloadConnections() error {
...
+func setupSnapSecurity(task *state.Task, snapInfo *snap.Info, repo *interfaces.Repository) error {
+	var snapState snapstate.SnapState
+	snapName := snapInfo.Name()
+	if err := snapstate.Get(task.State(), snapName, &snapState); err != nil {
+		task.Errorf("cannot get state of snap %q: %s", snapName, err)
+		return err
+	}
+	for _, backend := range securityBackends {
+		if err := backend.Setup(snapInfo, snapState.DevMode, repo); err != nil {
+			task.Errorf("cannot setup security of snap %q (backend %s): %s", snapName, backend.Name(), err)
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: `("cannot setup %s for snap %q: %s", backend.Name(), snapName, err)`

Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done, thanks for the suggestion :-)

**Comment by niemeyer** [comment]
Int tests infra is broken. Merging based on unit tests.

**Pull request merged**